### PR TITLE
feat: support projesetset for collection

### PR DESCRIPTION
**Pull request by Yisaer**

### CODE REVIEW

File: internal/topo/operator/projectset_operator.go [hunk 1]
```diff
@@ -32,34 +32,57 @@
...
-		if !ok {
-			return fmt.Errorf("can't find the result from the %v function", srfName)
+		newTuples, err := ps.handleSRFRow(srfName, input)
+		if err != nil {
+			return err
 		}
-		aValues, ok := aValue.([]interface{})
-		if !ok {
-			return fmt.Errorf("the result from the %v function should be array", srfName)
+		for _, tuple := range newTuples {
```
Comment (ngjaying) [inline internal/topo/operator/projectset_operator.go:44]: why need to copy the result?

File: internal/xsql/collection.go [hunk 2]
```diff
@@ -157,6 +158,22 @@
...
 
+func (w *WindowTuples) RangeProjectSet(f func(i int, r CloneAbleRow) ([]CloneAbleRow, error)) ([]Collection, error) {
+	newCollections := make([]Collection, 0)
+	for i, oldTupleRow := range w.Content {
+		newTuples, err := f(i, oldTupleRow.Clone())
+		if err != nil {
+			return nil, err
+		}
+		for _, newTuple := range newTuples {
+			ns := w.Clone().(*WindowTuples)
```
Comment (ngjaying) [inline internal/xsql/collection.go:169]: why return multiple collections? we can just flatten to a collection. It is also save to replace the old content

File: internal/xsql/collection.go [hunk 3]
```diff
@@ -37,6 +37,7 @@
 // Collection A collection of rows as a table. It is used for window, join, group by, etc.
 type Collection interface {
 	SortingData
+	RangeProjectSet(f func(i int, r CloneAbleRow) ([]CloneAbleRow, error)) ([]Collection, error)
```
Comment (ngjaying) [inline internal/xsql/collection.go:40]: This seems like only projectset needs, I don't think we should put it here. This interface should only provide some basic/atomic functions.

File: internal/topo/topotest/rule_test.go [hunk 4]
```diff
@@ -22,6 +22,207 @@ import (
 	"github.com/lf-edge/ekuiper/pkg/api"
 )
 
+func TestSRFSQL(t *testing.T) {
```
Comment (ngjaying) [inline internal/topo/topotest/rule_test.go:25]: Maybe add most tests in the operator level? Here is more expensive.

**Review by ngjaying — CHANGES_REQUESTED** [review]
Suggest to add more comments ~

File: internal/topo/topotest/rule_test.go (hunk 4 above)
Comment (Yisaer) [reply internal/topo/topotest/rule_test.go:25]: it's too hard to build testcase in operator level for this case.

File: internal/topo/operator/projectset_operator.go [hunk 5]
```diff
@@ -23,43 +23,101 @@ import (
...
-				newTupleRow.Set(srfName, v)
+		switch ts := input.(type) {
+		case *xsql.JoinTuples:
+			newTuples := make([]*xsql.JoinTuple, 0)
+			for _, tuple := range ps.collectionRows {
+				newTuples = append(newTuples, tuple.(*xsql.JoinTuple))
+			}
+			ts.Content = newTuples
+		case *xsql.GroupedTuplesSet:
+			newTuples := make([]*xsql.GroupedTuples, 0)
```
Comment (ngjaying) [inline internal/topo/operator/projectset_operator.go:72]: 0 -》len(ps.collectionRows)

Comment (ngjaying) [reply internal/topo/operator/projectset_operator.go:72]: Or make([]*xsql.GroupedTuples, 0, len(ps.collectionRows)) Or is it possible to have no new allocation/copy

Comment (ngjaying) [reply internal/topo/operator/projectset_operator.go:72]: Other types also have this problem

**Review by ngjaying — CHANGES_REQUESTED** [review]

File: internal/topo/operator/projectset_operator.go (hunk 5 above)
Comment (Yisaer) [reply internal/topo/operator/projectset_operator.go]: updated.

**Review by ngjaying — APPROVED** [review]

**Pull request merged**